ctapipe_io_magic/__init__.py

```diff
@@ -656,63 +751,120 @@ def parse_simulation_header(self):
+                min_az = run_header_tree['MMcRunHeader_1.fShowerPhiMin'].array(library="np")[0]
+                max_wavelength = run_header_tree['MMcRunHeader_1.fCWaveUpper'].array(library="np")[0]
+                min_wavelength = run_header_tree['MMcRunHeader_1.fCWaveLower'].array(library="np")[0]
+
+            simulation_config[run_number] = SimulationConfigContainer(
+                corsika_version=corsika_version,
+                energy_range_min=u.Quantity(e_low, u.GeV).to(u.TeV),
+                energy_range_max=u.Quantity(e_high, u.GeV).to(u.TeV),
+                prod_site_alt=u.Quantity(site_height, u.cm).to(u.m),
+                spectral_index=spectral_index,
+                num_showers=n_showers,
+                shower_reuse=1,
+                # shower_reuse not written in the magic root file, but since the
+                # sim_events already include shower reuse we artificially set it
+                # to 1 (actually every shower reused 5 times for std MAGIC MC)
+                shower_prog_id=1,
+                prod_site_B_total=MAGIC_Btot,
+                prod_site_B_declination=MAGIC_Bdec,
+                prod_site_B_inclination=MAGIC_Binc,
+                max_alt=u.Quantity((90. - min_zd), u.deg).to(u.rad),
+                min_alt=u.Quantity((90. - max_zd), u.deg).to(u.rad),
+                max_az=u.Quantity(max_az, u.deg).to(u.rad),
+                min_az=u.Quantity(min_az, u.deg).to(u.rad),
+                max_viewcone_radius=view_cone * u.deg,
+                min_viewcone_radius=0.0 * u.deg,
+                max_scatter_range=u.Quantity(max_impact, u.cm).to(u.m),
+                min_scatter_range=0.0 * u.m,
+                atmosphere=atm_model,
+                corsika_wlen_min=min_wavelength * u.nm,
+                corsika_wlen_max=max_wavelength * u.nm,
+            )
+
+        return simulation_config
+
+    def prepare_drive_information(self):
+
+        drive_array_list = [
+            'MReportDrive.fMjd',
+            'MReportDrive.fCurrentZd',
+            'MReportDrive.fCurrentAz',
+            'MReportDrive.fRa',
+            'MReportDrive.fDec'
+        ]
+
+        drive_data = dict()
+        drive_data['mjd'] = np.array([])
+        drive_data['zd'] = np.array([])
+        drive_data['az'] = np.array([])
+        drive_data['ra'] = np.array([])
+        drive_data['dec'] = np.array([])
+
+        # Getting the telescope drive info
+        for rootf in self.files_:
+            drive = rootf["Drive"].arrays(drive_array_list, library="np")
+
+            drive_mjd = drive['MReportDrive.fMjd']
+            drive_zd = drive['MReportDrive.fCurrentZd']
+            drive_az = drive['MReportDrive.fCurrentAz']
+            drive_ra = drive['MReportDrive.fRa'] * degrees_per_hour
+            drive_dec = drive['MReportDrive.fDec']
+
+            drive_data['mjd'] = np.concatenate((drive_data['mjd'], drive_mjd))
```

Comment:
Appending / concatenating numpy arrays over and over is very inefficent. It is better to use a python list and concatenate once in the end:

```
data = []

for f in file:
    data.append(new_data)

data = np.concatenate(data)
```

Reply:
This could probably written better and more concise as (not tested):

```
    def prepare_drive_information(self):

        drive_leaves = {
            'mjd': 'MReportDrive.fMjd',
            'zd': 'MReportDrive.fCurrentZd',
            'az': 'MReportDrive.fCurrentAz',
            'ra': 'MReportDrive.fRa',
            'dec': 'MReportDrive.fDec',
        }


        # Getting the telescope drive info
        drive_data = []
        for rootf in self.files_:
            drive = rootf["Drive"].arrays(drive_leaves.values(), library="np")

            if len(drive) < 3:
                LOGGER.warning(f"File {rootf.file_path} has only {len(drive)} drive reports.")
                if len(drive) == 0:
                    raise MissingDriveReportError(f"File {rootf.file_path} does not have any drive report. Check if it was merpped correctly.")

            drive_data.append(drive)

        drive_data = np.concatenate(drive_data)

        # rename field names
        drive_data.dtype.names = list(drive_leaves.keys())

        # convert unit as before (Better use astropy units!)
        drive_data['ra'] *= degrees_per_hour


        # get only drive reports with unique times, otherwise interpolation fails.
        _, unique_indices = np.unique(
            drive_data['mjd'],
            return_index=True
        )

        return drive_data[unique_indices]

```

Reply:
Does not seem to work, it gives the error:

```
ValueError: zero-dimensional arrays cannot be concatenated
```

I guess the problem is that `drive` is a dict where the keys are `drive_leaves.values()` and the values are `np.ndarray`s, while `np.concatenate` expects an array_like as input. 

Looking a bit in uproot, there is a concatenate function which can be used like this:

```python
uproot_files = [ {rootf.file_path : "Drive"} for rootf in self.files_]
drive_data = uproot.concatenate(files=uproot_files, expressions=drive_array_list, library="np")
```

Which returns directly a dictionary with the final concatenated arrays. The problem I have with this approach is that the name of the keys are the ones in `drive_array_list` (but I could do `drive_data[k_new] = drive_data.pop(k_old)` to rename them, of course it can be done in a one liner for all the keys) and that I cannot check the number of drive reports for each file.

Reply:
> I guess the problem is that drive is a dict where the keys are drive_leaves.values() and the values are np.ndarrays, while np.concatenate expects an array_like as input.

Ah, sorry, my bad. I thought the return value would be a numpy structured array, not a dict.

Reply:
This should work then:

```
    def prepare_drive_information(self):

        drive_leaves = {
            'mjd': 'MReportDrive.fMjd',
            'zd': 'MReportDrive.fCurrentZd',
            'az': 'MReportDrive.fCurrentAz',
            'ra': 'MReportDrive.fRa',
            'dec': 'MReportDrive.fDec',
        }


        # Getting the telescope drive info
        drive_data = {k: [] for k in drive_leaves.keys()}

        for rootf in self.files_:
            drive = rootf["Drive"].arrays(drive_leaves.values(), library="np")

            n_reports = len(drive['MReportDrive.fMjd'])
            if n_reports == 0:
                raise MissingDriveReportError(f"File {rootf.file_path} does not have any drive report. Check if it was merpped correctly.")
            elif n_reports < 3:
                LOGGER.warning(f"File {rootf.file_path} has only {n_reports} drive reports.")

            for key, leaf in drive_leaves.items():
                drive_data[key].append(drive[leaf])

        drive_data = {
            k: np.concatenate(v)
            for k, v in drive_data.items()
        }

        # convert unit as before (Better use astropy units!)
        drive_data['ra'] *= degrees_per_hour


        # get only drive reports with unique times, otherwise interpolation fails.
        _, unique_indices = np.unique(
            drive_data['mjd'],
            return_index=True
        )
        drive_data = {
            k: v[unique_indices]
            for k, v in drive_data.items()
        }
        return drive_data
```
```

Reply:
Indeed, it works nicely.